[숫자 야구 게임] 이지윤 과제 제출합니다.

dd-jiyun opened the pull request:
### :boom: 어려웠던 점
- 인공지능의 도움과 다른 사람의 코드를 보지 않고 스스로 코드를 짜려고 하니 생각보다 막막했습니다.
- 페어프로그래밍을 하면서 기능 단위로 역할을 분담해서 각자의 스타일로 코드를 구성하려고 하니 다른 의견을 가지고 있을 때 설명하는 것에 어려움을 겪었습니다.

### :monocle_face: 궁금한 점
- 클래스를 나눌 때 한 기능을 하게 하는 게 객체 지향적이라고 알고 있는데 지금 나눈거에서 더 나눴으면 하는 게 있는지 궁금합니다.
- 패키지 분리를 어떻게 하는게 좋은지 감이 안잡혔습니다..

hangillee committed:
Update README.md

hangillee committed:
Update README.md

sojeong0202 committed:
:sparkles: [feat] 난수 생성 기능 구현

dd-jiyun committed:
:sparkles: [feat] 사용자에게 숫자 입력받기

sojeong0202 committed:
:sparkles: [feat] 결과 출력 및 게임 실행 로직 구현

- 사용자 입력 시 [스트라이크, 볼, 낫싱] 판단 후 출력
- 사용자에게 재시작 여부 입력 받기
- 게임 프로세스 로직 완성

sojeong0202 committed:
:sparkles: [feat] 결과 출력 및 게임 실행 로직 구현

dd-jiyun committed:
:sparkles: [refactor] 전반적인 코드 리팩토링

- IllegalArgumentException 예외를 ExceptionHandler로 메세지를 출력할 수 있게 하였습니다.
- StringBuilder로 버퍼를 만들어 해당 조건에 맞는 출력문을 추가하여 결과를 반환하게 하였습니다.
- 스트라이크, 볼 개수를 세는 과정에서 서로 의존하지 않게 수정하였습니다.
- for문을 사용하여 난수와 사용자가 입력한 값이 같은 인덱스에 있고, 같은 수인지 판단하여 스트라이크 수를 계산하였습니다.
- 사용하지 않는 import문을 지우고, static으로 선언하여 코드를 간결하게 만들었습니다.

dd-jiyun committed:
:memo: [docs] 구현 기능 및 느낀점

jiyoungmerong commented on docs/README.md:
```diff
@@ -0,0 +1,69 @@
+- StringBulider로 새로운 객체를 생성하여 문자열 조합하는 방법에 대해서 알게 되었습니다.
+
+### **[느낀 점]**
+- 역시 지윤 언니... 그 이름의 명성에 대해서 알게 되었습니다. 혼자할 때, 모호하고 어렵기만 했던 점이 함께 하니(거의 언니가 캐리) 
```
README도 프로그램의 일부분이기 때문에 개인적인 메모는 최대한 지양하는 것이 좋습니다ㅎㅎ

jiyoungmerong commented on src/main/java/baseball/model/Player.java:
```diff
@@ -0,0 +1,29 @@
+
+    public List<Integer> inputPlayerNumber() {
+        printInputMessage();
+        String input = Console.readLine();
```
Model, View, Controller에서 입력의 책임을 가지고 있는 단은 어디일까요?!

jiyoungmerong commented on src/main/java/baseball/model/Player.java:
```diff
@@ -0,0 +1,29 @@
+public class Player {
+
+    public List<Integer> inputPlayerNumber() {
+        printInputMessage();
```
Player 클래스는 모델 영역으로 보여요!
모델 영역에서 입력 메세지를 출력하는 것이 바람직할까요?
어느 영역의 책임일까요?

jiyoungmerong commented on src/main/java/baseball/model/Player.java:
```diff
@@ -0,0 +1,29 @@
+        return numbers;
+    }
+
+    private List<Integer> convertStringToIntegers(final String input) {
+        return new ArrayList<>(input.chars()
+                .map(Character::getNumericValue)
+                .boxed()
+                .toList());
+    }
```
Stream 활용 좋아요! 문자열을 정수 리스트로 반환하는 메소드같네요. 
여러 곳에서 사용할 수 있을 것 같은데, 유틸리티 클래스가 아닌 Player 클래스 내에 작성한 이유가 있을까요?!

jiyoungmerong commented on src/main/java/baseball/service/BaseballGame.java:
```diff
@@ -0,0 +1,61 @@
+
+public class BaseballGame {
+
+    private static final int RESTART_YES = 1;
+    private static final int RESTART_NO = 2;
```
상수 사용 좋아요!
상수 사용으로 매직 넘버를 대체하면 어떤 이점이 있을까요?!

dd-jiyun replied:
가독성이 높아진다는 이점이 있는 걸로 알고 있습니다!
혹시 또 다른 이점이 있을까요?

jiyoungmerong replied:
가독성이 높아지는게 엄청 큰 이점같아요
저는 상수를 사용하면 해당 값의 의미나 역할을 명확하게 전달해줘서 좋더라구요!!
그래서 유지보수 면에서도 좋은 것 같습니다!

jiyoungmerong commented on src/main/java/baseball/service/BaseballGame.java:
```diff
@@ -0,0 +1,61 @@
+        this.countResult = countResult;
+    }
+
+    public void playGame() {
+
+        do {
+            printStartMessage();
+            processGame();
+        } while (askRestartGame());
+    }
+
+    public void processGame() {
+        List<Integer> computerNumber = computer.generateRandomNumber();
+        List<Integer> playerNumber;
+        do {
+            playerNumber = player.inputPlayerNumber();
+
+            int countStrikes = countResult.countStrikes(playerNumber, computerNumber);
+            int countBalls = countResult.countBalls(playerNumber, computerNumber);
+
+            printResultMessage(countStrikes, countBalls);
+        } while (!countResult.judgementEndGame(playerNumber, computerNumber));
+
+        printEndGameMessage();
+    }
+
+    public boolean askRestartGame() {
+        printRestartGameMessage();
+        int answerNumber = Integer.parseInt(Console.readLine());
+        if (answerNumber != RESTART_YES && answerNumber != RESTART_NO) {
+            Exception(RESTART_INPUT_WRONG_NUMBER);
+        }
+        return answerNumber == RESTART_YES;
+    }
```
컨트롤러는 없고 서비스단만 존재하네요! 
게임 시작, 진행, 재시작 등은 비지니스 로직을 호출하여 진행된다고 생각해서,
컨트롤러단에서 진행되는게 맞다고 생각하는데
혹시 서비스단에서 진행하는 이유가 있을까요?

jiyoungmerong commented on src/main/java/baseball/service/CountResult.java:
```diff
@@ -0,0 +1,34 @@
+
+    public int countStrikes(List<Integer> playerNumbers, List<Integer> computerNumbers) {
+        int strikes = 0;
+        for (int i = 0; i < playerNumbers.size(); i++) {
+            if (playerNumbers.get(i).equals(computerNumbers.get(i))) {
+                strikes++;
+            }
```
기회가 된다면 반복문 대신 Stream을 사용해서 더 간결하게 표현해볼 수 있을 것 같아요!

Integer 객체가 주어진다는 보장이 있는데 왜 equals()를 사용했는지 궁금해요!

jiyoungmerong commented on src/main/java/baseball/service/CountResult.java:
```diff
@@ -0,0 +1,34 @@
+        return strikes;
+    }
+
+    public int countBalls(List<Integer> playerNumbers, List<Integer> computerNumbers) {
+        int balls = 0;
+        for (Integer number : playerNumbers) {
+            if (computerNumbers.contains(number) && !playerNumbers.get(playerNumbers.indexOf(number))
+                    .equals(computerNumbers.get(playerNumbers.indexOf(number)))) {
+                balls++;
+            }
+        }
+        return balls;
+    }
```
`if (computerNumbers.contains(number) && !playerNumbers.get(playerNumbers.indexOf(number)).equals(computerNumbers.get(playerNumbers.indexOf(number))))`
조건문이 가독성이 떨어지는 것 같아요.
Ball 인지 확인하는 메소드를 만들어서 사용하거나, 각각의 메소드로 분리하는 게 좋아보여요!

jiyoungmerong commented on src/main/java/baseball/util/ErrorMessage.java:
```diff
@@ -0,0 +1,15 @@
+package baseball.util;
+
+public enum ErrorMessage {
```
util 패키지에 enum 사용 좋아요!

jiyoungmerong commented on src/main/java/baseball/util/ExceptionHandler.java:
```diff
@@ -0,0 +1,9 @@
+
+    public static void Exception(ErrorMessage errorMessage) {
+        System.out.println("[Error] " + errorMessage.message);
+        throw new IllegalArgumentException(errorMessage.message);
```
만약 null이 들어온다면 해당 Exception에서 처리가 가능할까요?

dd-jiyun replied:
```java
 if (input == null) {
        Exception(INPUT_NUMBER_NOT_NULL);
        }
```
IllegalArgumentException이 잘못된 인자를 넘겨줬을 때 예외가 발생하는 거라서 null도 포함된다고 생각하여 위와 같이 조건문으로 다뤄봤습니다!

jiyoungmerong commented on src/main/java/baseball/util/ExceptionHandler.java:
```diff
@@ -0,0 +1,9 @@
+
+public class ExceptionHandler {
+
+    public static void Exception(ErrorMessage errorMessage) {
```
Exception을 메소드명으로 작성해도 괜찮을까요?!

jiyoungmerong commented on src/main/java/baseball/util/Validator.java:
```diff
@@ -0,0 +1,47 @@
+
+    private static final int INPUT_NUMBER_LENGTH = 3;
+
+    public static void validateInputNumber(String input) {
+        // null 값 검증
+        if (input == null) {
+            Exception(INPUT_NUMBER_NOT_NULL);
+        }
+        // 3자리 검증
+        if (input.length() != INPUT_NUMBER_LENGTH) {
+            Exception(INPUT_NUMBER_NOT_3_DIGIT);
+        }
+        // 중복된 수 검증
+        if (hasDuplicateNumber(input)) {
+            Exception(INPUT_NUMBER_DUPLICATE);
+        }
+        // 요구조건 검증
+        if (!isRequirementNumber(input)) {
+            Exception(INPUT_NUMBER_NOT_RANGE);
+        }
+
+    }
```
validateInpurNumber 메소드에서 여러 유효성 검사를 진행하고 있어요.
검사를 수행하는 메소드를 각각 분리하여, 하나의 책임을 가질 수 있도록 수정하는 건 어떨까요?

하나의 메소드에 작성하는 것과 각각의 메소드로 분리하는 것은 어떤 차이가 있을까요?

jiyoungmerong commented on src/main/java/baseball/util/Validator.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    private static boolean isRequirementNumber(String input) {
+        return Pattern.matches("^[1-9]{3}$", input);
```
정규식 좋아요!

jiyoungmerong commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,46 @@
+        System.out.println("게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.");
+    }
+
+    /**
+     * 볼, 스트라이크, 낫싱을 출력한다.
+     *
+     * @param strikes 스트라이크 수
+     * @param balls   볼 수
+     */
```
저는 이런 식으로 주석 작성하는 거 너무 좋아합니다!!
물론 모든 메소드마다 작성할 필요는 없다고 생각하지만, 좋은 습관 같습니다!

jiyoungmerong reviewed (commented):
1. 클래스를 나눌 때 한 기능을 하게 하는 게 객체 지향적이라고 알고 있는데 지금 나눈거에서 더 나눴으면 하는 게 있는지 궁금합니다.
=> MVC에서 Controller가 하는 역할이 뭔지 한 번 고민해보세요!

2. 패키지 분리를 어떻게 하는게 좋은지 감이 안잡혔습니다..
=> 패키지 분리는 잘 하신 것 같아요. 다만 해당 패키지 내에서 역할 분담을 생각해서 리팩토링 하시면 될 것 같아요! 서비스단에서 진행되는 로직을 컨트롤러의 역할을 생각하여 분리해보세요!

dd-jiyun committed:
refactor: 숫자 야구 게임

- Player는 숫자를 받아서 검증하고 변환하는 역할을 합니다.
- 각 역할에 맞는 패키지로 옮겼습니다.
- interface로 추상화를 하여 class에 implements하여 구현하는 방식을 사용했습니다.
- Exception 메소드 명을 CustomException으로 수정하였습니다.
- Validator 클래스에서 검증하는 부분을 메서드를 분리하였습니다.

dd-jiyun commented:
@jiyoungmerong 리팩토링 했습니다! 코드 리뷰 부탁드립니다 🙇🏻‍♀️

jiyoungmerong commented on src/main/java/baseball/util/Validator.java:
```diff
@@ -0,0 +1,58 @@
+
+    private static final int INPUT_NUMBER_LENGTH = 3;
+
+    private static void checkedNullInputNumber(String input) {
+        if (input == null) {
+            CustomException(INPUT_NUMBER_NOT_NULL);
+        }
+    }
+
+    private static void checkedLengthInputNumber(String input) {
+        if (input.length() != INPUT_NUMBER_LENGTH) {
+            CustomException(INPUT_NUMBER_NOT_3_DIGIT);
+        }
+    }
```
메소드로 분리 좋네요!

jiyoungmerong commented on src/main/java/baseball/util/DefaultNumberConverter.java:
```diff
@@ -0,0 +1,15 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class DefaultNumberConverter implements NumberConverter {
+    @Override
```
의존성 부분에서 어떤 장점이 있을까요!

jiyoungmerong commented on src/main/java/baseball/model/Player.java:
```diff
@@ -0,0 +1,25 @@
+import baseball.util.NumberConverter;
+import java.util.List;
+
+public class Player {
+
+    private final InputReader inputReader;
+    private final NumberConverter numberConverter;
+
+    public Player(InputReader inputReader, NumberConverter numberConverter) {
+        this.inputReader = inputReader;
+        this.numberConverter = numberConverter;
+    }
+
+    public List<Integer> inputPlayerNumber() {
+        String input = inputReader.readInput();
+        validateInputNumber(input);
+
+        return numberConverter.convertStringToIntegers(input);
+    }
```
전반적으로 Player 모델의 코드가 많이 깔끔해진 것 같아요!

jiyoungmerong commented on src/main/java/baseball/util/ExceptionHandler.java:
```diff
@@ -0,0 +1,9 @@
+
+public class ExceptionHandler {
+
+    public static void CustomException(ErrorMessage errorMessage) {
```
Exception -> CustomException 명시적인 표현 좋아요!

jiyoungmerong commented on src/main/java/baseball/service/BaseballGame.java:
```diff
@@ -0,0 +1,73 @@
+        this.countResult = countResult;
+    }
+
+    public void playGame() {
+        do {
+            printStartMessage();
+            processGame();
+        } while (askRestartGame());
+    }
```
게임 시작이나 진행 등은 제가 생각할 땐 컨트롤러의 역할에 더 맞는 것 같아요!

jiyoungmerong commented on src/main/java/baseball/service/CountResult.java:
```diff
@@ -0,0 +1,38 @@
+
+        for (Integer number : playerNumbers) {
+            if (computerNumbers.contains(number) && !playerNumbers.get(playerNumbers.indexOf(number))
+                    .equals(computerNumbers.get(playerNumbers.indexOf(number)))) {
```
조건문이 아직 복잡한 것 같아요!
메소드로 분리하면 더 쉬울 것 같아요.

jiyoungmerong reviewed (commented):
코드가 많이 깔끔해진 것 같아요! 인터페이스 사용도 좋은 것 같습니다!
복잡한 로직의 경우 한번에 적으려 하지 말고, 메소드를 사용하면 더 좋은 코드가 될 수 있을 것 같아요!
